lib/ansible/modules/crypto/openssl_dhparam.py

```diff
@@ -0,0 +1,252 @@
+'''
+
+EXAMPLES = '''
+# Generate Diffie-Hellman parameters with the default size (4096 bits)
+- openssl_dhparam:
+    path: /etc/ssl/dhparams.pem
+
+# Generate DH Parameters with a different size (2048 bits)
+- openssl_dhparam:
+    path: /etc/ssl/dhparams.pem
+    size: 2048
+
+# Force regenerate an DH parameters if they already exist
+- openssl_dhparam:
+    path: /etc/ssl/dhparams.pem
+    force: True
+
+'''
+
+RETURN = '''
+size:
+    description: Size (in bits) of the Diffie-Hellman parameters
+    returned: changed or success
+    type: int
+    sample: 4096
+filename:
+    description: Path to the generated Diffie-Hellman parameters
+    returned: changed or success
+    type: string
+    sample: /etc/ssl/dhparams.pem
+'''
+
+import os
+import re
+
+"""Note: We also monkey-patch subprocess for python 2.6 to
+give feature parity with later versions.
+Monkey-patch from http://pydoc.net/pep8radius/0.9.0/ with slight modifications
+MIT Licence
+
+Remove patch when Python 2.6 is no longer supported
+"""
+import subprocess
+try:
+    from subprocess import check_output, CalledProcessError
+except ImportError:  # pragma: no cover
+    # python 2.6 doesn't include check_output
+    # monkey patch it in!
+    def check_output(*popenargs, **kwargs):
+        if 'stdout' in kwargs:  # pragma: no cover
+            raise ValueError('stdout argument not allowed, '
+                             'it will be overridden.')
+        process = subprocess.Popen(stdout=subprocess.PIPE,
+                                   *popenargs, **kwargs)
+        output = process.communicate()[0]
+        retcode = process.poll()
+        if retcode:
+            cmd = kwargs.get("args")
+            if cmd is None:
+                cmd = popenargs[0]
+            raise CalledProcessError(retcode, cmd, output=output)
+        return output
+
+    # overwrite CalledProcessError due to `output`
+    # keyword not being available (in 2.6)
+    class CalledProcessError(Exception):
+        """Exception raised for call errors"""
+
+        def __init__(self, returncode, cmd, output=None):
+            self.returncode = returncode
+            self.cmd = cmd
+            self.output = output
+
+        def __str__(self):
+            return "Command '%s' returned non-zero exit status %d" % (
+                self.cmd, self.returncode)
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
+
+
+class DHParameterError(Exception):
+    pass
+
+
+class DHParameter(object):
+
+    def __init__(self, module):
+        self.state = module.params['state']
+        self.path = module.params['path']
+        self.size = int(module.params['size'])
+        self.force = module.params['force']
+        self.changed = False
+
+    def generate(self, module):
+        """Generate a keypair."""
+        # openssl dhparam -out <path> <bits>
+        if not self.force and self.check(module):
+            return
```

Comment:
I suspect there is a subtle bug lurking here. If I generate a dhparam file, and then later changes permissions, the return will prevent the permission to be applied, since that's done at the end of the function and only if we did generate a param.

Reply:
`check()` will return `False` is the attributes (permissions) do not match; this will make it pass this check and then change the permissions (and re-generate parameters in this case – perhaps I should refactor it to only do that if the bits don't match).